### PR TITLE
Implementation for Close buttons on tabs

### PR DESCRIPTION
**Pull request by voidburn**
This is a proposal for close buttons on tabs. It should be backwards compatible (defaulting to no buttons).

# Notes 
- I had to update dependencies, so this PR includes a new cargo.lock with Egui 0.27.2

### CODE REVIEW

**Comment by voidburn** [comment]
P.S. In my personal implementation I commented out all `.on_hover_cursor(egui::CursorIcon::Grab)` except for the one that changes the cursor while dragging, as they were overriding all my attempts to change the cursor to a pointer when over the close button. I don't know how that would've impacted existing implementations so I left them in this PR.

Cursor changes in question: 
- https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs#L279
- https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs#L307


As a result, the mouse cursor remains the drag hand even when hovering the close button.

File: src/behavior.rs [hunk 1]
```diff
@@ -22,17 +22,37 @@ pub enum EditAction {
...
     /// You can make the pane draggable by returning [`UiResponse::DragStarted`]
     /// when the user drags some handle.
-    fn pane_ui(&mut self, _ui: &mut Ui, _tile_id: TileId, _pane: &mut Pane) -> UiResponse;
+    fn pane_ui(&mut self, ui: &mut Ui, tile_id: TileId, pane: &mut Pane) -> UiResponse;
 
     /// The title of a pane tab.
     fn tab_title_for_pane(&mut self, pane: &Pane) -> WidgetText;
 
+    /// Should the tab be closable?
+    fn closable(&self) -> bool {
```
Comment (emilk) [inline src/behavior.rs:45]:
```suggestion
    /// Should tabs have a close-button?
    ///
    /// If `true` and the button is clicked, [`Self::on_tab_close`] is called.
    fn closable(&self) -> bool {
```

File: src/behavior.rs [hunk 2]
```diff
@@ -22,17 +22,37 @@ pub enum EditAction {
...
+        false
+    }
+
+    /// If `closable()` returns true, hat should be done when the close button is pressed?
```
Comment (emilk) [inline src/behavior.rs:49]:
```suggestion
    /// If [`Self::closable`] returns true, hat should be done when the close button is pressed?
```

File: src/behavior.rs [hunk 3]
```diff
@@ -22,17 +22,37 @@ pub enum EditAction {
...
+    ///
+    /// Return `true` if the tab should be closed, `false` otherwise.
+    fn on_tab_close(&mut self, _tiles: &mut Tiles<Pane>, _tile_id: TileId) -> bool {
+        false
+    }
```
Comment (emilk) [inline src/behavior.rs:54]: We should consider removing this and instead have

```suggestion
    /// Should the tab have a close-button?
	fn is_tab_closable(&self, tiles: &Tiles<Pane>, tile_id: TileId) -> bool {
	    false
	}    
	
	/// Called after a tab has been closed.
	fn on_tab_close(&mut self, _tiles: &mut Tiles<Pane>, _tile_id: TileId) {
    }
```

That way the user has exact control over which tabs should have a close-button.

File: src/behavior.rs [hunk 4]
```diff
@@ -48,6 +68,18 @@ pub trait Behavior<Pane> {
...
+    fn pane<'a>(&'a mut self, tiles: &'a Tiles<Pane>, tile_id: TileId) -> Option<&'a Pane> {
+        if let Some(tile) = tiles.get(tile_id) {
+            match tile {
+                Tile::Pane(pane) => Some(pane),
+                Tile::Container(_) => None,
+            }
+        } else {
+            None
+        }
+    }
```
Comment (emilk) [inline src/behavior.rs:81]: This looks like it should be a helper member of `Tiles` instead?

File: src/behavior.rs [hunk 5]
```diff
@@ -22,17 +22,37 @@ pub enum EditAction {
     TabSelected,
 }
 
+/// The state of a tab, used to inform the rendering of the tab.
+#[derive(Clone, Debug, Default)]
+pub struct TabState {
+    pub active: bool,
+    pub is_being_dragged: bool,
+    pub closable: bool,
```
Comment (emilk) [inline src/behavior.rs:30]: Please add docstrings!

File: src/behavior.rs [hunk 6]
```diff
@@ -59,50 +91,108 @@ pub trait Behavior<Pane> {
...
         let font_id = TextStyle::Button.resolve(ui.style());
         let galley = text.into_galley(ui, Some(false), f32::INFINITY, font_id);
 
         let x_margin = self.tab_title_spacing(ui.visuals());
-        let (_, rect) = ui.allocate_space(vec2(
-            galley.size().x + 2.0 * x_margin,
-            ui.available_height(),
-        ));
-        let response = ui.interact(rect, id, Sense::click_and_drag());
+        let (_, tab_rect) = match state.closable {
```
Comment (emilk) [inline src/behavior.rs:107]: use `let width = if closeable … else …;` instead

File: src/behavior.rs [hunk 7]
```diff
@@ -59,50 +91,108 @@ pub trait Behavior<Pane> {
...
+        let tab_response = ui.interact(tab_rect, id, Sense::click_and_drag());
 
         // Show a gap when dragged
-        if ui.is_rect_visible(rect) && !is_being_dragged {
-            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, active);
-            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, active);
-            ui.painter().rect(rect.shrink(0.5), 0.0, bg_color, stroke);
+        if ui.is_rect_visible(tab_rect) && !state.is_being_dragged {
+            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, state.active);
+            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, state.active);
```
Comment (emilk) [inline src/behavior.rs:123]: let's pass the whole state in to these functions

```suggestion
            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, state);
            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, state);
```

File: src/behavior.rs [hunk 8]
```diff
@@ -59,50 +91,108 @@ pub trait Behavior<Pane> {
...
+        tiles: &mut Tiles<Pane>,
         ui: &mut Ui,
         id: Id,
         tile_id: TileId,
-        active: bool,
-        is_being_dragged: bool,
+        state: TabState,
     ) -> Response {
         let text = self.tab_title_for_tile(tiles, tile_id);
+        let close_btn_size = Vec2::splat(ui.spacing().icon_width);
```
Comment (emilk) [inline src/behavior.rs:101]: You use `icon_width` but then call `.shrink(2.0)` on it later on, so it isn't really `icon_width`.

Just make it configureable with `Behavior::close_button_size` instead

**Review by emilk — COMMENTED** [review]
Thanks, this looks promising!

**Comment by voidburn** [comment]
Thanks for the comments, I'll get to it during the week or next weekend! Everything looks very reasonable!

File: src/behavior.rs (hunk 3 above)
Comment (voidburn) [reply src/behavior.rs]: Agreed on the `closable` signature change, I do require the bool returning from `on_tab_close`, for example an implementation might want to prompt the user to save changes (or cancel closing the panel). That's what gives the user to cancel the close request.

File: Cargo.lock [hunk 9]
```diff
@@ -4,9 +4,9 @@ version = 3
 
 [[package]]
 name = "ab_glyph"
-version = "0.2.23"
```
Comment (emilk) [inline Cargo.lock:7]: Please revert all changes to `Cargo.lock` (`git checkout main Cargo.lock`)

File: src/behavior.rs [hunk 10]
```diff
@@ -22,17 +22,43 @@ pub enum EditAction {
...
 
+    /// Should the tab have a close-button?
+    fn is_tab_closable(&self, _tiles: &Tiles<Pane>, _tile_id: TileId) -> bool {
+        false
+    }
+
+    /// Has the user confirmed that a tab should be closed?
+    fn on_tab_close(&mut self, _tiles: &mut Tiles<Pane>, _tile_id: TileId) -> bool {
+        false
+    }
```
Comment (emilk) [inline src/behavior.rs:55]:
```suggestion
    /// Called when the close-button on a tab is pressed.
    ///
    /// Return `false` to abort the closing of a tab (e.g. after showing a message box).
    fn on_tab_close(&mut self, _tiles: &mut Tiles<Pane>, _tile_id: TileId) -> bool {
        true
    }
```

File: src/behavior.rs [hunk 11]
```diff
@@ -59,50 +85,109 @@ pub trait Behavior<Pane> {
...
+            ui.allocate_space(vec2(
+                galley.size().x + close_btn_size.x + close_btn_left_padding + (2.0 * x_margin),
+                ui.available_height(),
+            ))
+        } else {
+            ui.allocate_space(vec2(
+                galley.size().x + (2.0 * x_margin),
+                ui.available_height(),
+            ))
+        };
```
Comment (emilk) [inline src/behavior.rs:111]:
```suggestion
        let button_width = galley.size().x + 2.0 * x_margin
            + f32::from(state.closable) * (close_btn_left_padding + close_btn_size.x);
        let (_, tab_rect) = ui.allocate_space(vec2(button_width, ui.available_height()));
```

File: src/behavior.rs [hunk 12]
```diff
@@ -59,50 +85,109 @@ pub trait Behavior<Pane> {
...
-                text_color,
-            );
+            // Prepare title's text for rendering
+            let text_color = self.tab_text_color(ui.visuals(), tiles, tile_id, state.active);
+            let mut text_position = egui::Align2::LEFT_CENTER
+                .align_size_within_rect(galley.size(), tab_rect)
+                .min;
+
+            // Respect the left margin
+            text_position.x += x_margin;
```
Comment (emilk) [inline src/behavior.rs:138]:
```suggestion
            let text_position = egui::Align2::LEFT_CENTER
                .align_size_within_rect(galley.size(), tab_rect.shrink(x_margin))
                .min;
```

File: src/behavior.rs [hunk 13]
```diff
@@ -59,50 +85,109 @@ pub trait Behavior<Pane> {
...
+            // Conditionally render the close button
+            if state.closable {
+                let close_btn_pad = (tab_rect.height() - close_btn_size.y) / 2.0;
+                let close_btn_rect = Rect::from_min_size(
+                    Pos2::from([
+                        tab_rect.right() - close_btn_pad - close_btn_size.x,
+                        tab_rect.center().y - 0.5 * close_btn_size.y,
+                    ]),
+                    close_btn_size,
+                );
```
Comment (emilk) [inline src/behavior.rs:152]:
```suggestion
                let close_btn_rect = egui::Align2::RIGHT_CENTER
                    .align_size_within_rect(close_btn_size, tab_rect.shrink(x_margin))
```

File: src/behavior.rs [hunk 14]
```diff
@@ -59,50 +85,109 @@ pub trait Behavior<Pane> {
...
+
+                // Allocate
+                let close_btn_id = ui.auto_id_with("tab_close_btn");
+                let close_btn_response = ui.interact(close_btn_rect, close_btn_id, Sense::click());
+
+                // Update the current UI's rect to include the close button
+                ui.expand_to_include_rect(close_btn_response.rect);
```
Comment (emilk) [inline src/behavior.rs:159]: This shouldn't be needed - `ui.allocate_space` has already allocated the superset `tab_rect`

File: src/behavior.rs [hunk 15]
```diff
@@ -59,50 +85,109 @@ pub trait Behavior<Pane> {
...
+
+                let visuals = ui.style().interact(&close_btn_response);
+
+                // Scale based on the interaction visuals
+                let rect = close_btn_rect.shrink(2.0).expand(visuals.expansion);
```
Comment (emilk) [inline src/behavior.rs:164]:
```suggestion
                let rect = close_btn_rect.expand(visuals.expansion);
```

File: src/tiles.rs [hunk 16]
```diff
@@ -122,6 +122,19 @@ impl<Pane> Tiles<Pane> {
         self.tiles.values_mut()
     }
 
+    /// Get the pane instance for a given [`TileId`]
+    #[allow(dead_code)]
+    fn pane(&self, tile_id: &TileId) -> Option<&Pane> {
```
Comment (emilk) [inline src/tiles.rs:127]: This belong next to `get` as it is a similar thing

File: src/tiles.rs [hunk 17]
```diff
@@ -122,6 +122,19 @@ impl<Pane> Tiles<Pane> {
...
+        if let Some(tile) = self.tiles.get(tile_id) {
+            match tile {
+                Tile::Pane(pane) => Some(pane),
+                Tile::Container(_) => None,
+            }
+        } else {
+            None
+        }
```
Comment (emilk) [inline src/tiles.rs:135]:
```suggestion
        match self.tiles.get(tile_id)? {
            Tile::Pane(pane) => Some(pane),
            Tile::Container(_) => None,
        }
```

**Review by emilk — CHANGES_REQUESTED** [review]

File: src/tiles.rs (hunk 16 above)
Comment (voidburn) [reply src/tiles.rs:127]: Took the liberty to also fix the visibility of the function (should be public) and rename it to `get_pane`

**Comment by emilk** [comment]
> P.S. In my personal implementation I commented out all `.on_hover_cursor(egui::CursorIcon::Grab)` except for the one that changes the cursor while dragging, as they were overriding all my attempts to change the cursor to a pointer when over the close button. I don't know how that would've impacted existing implementations so I left them in this PR.
> 
> Cursor changes in question:
> 
> * [`main`/src/container/tabs.rs#L279](https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs?rgh-link-date=2024-05-31T10%3A45%3A05Z#L279)
> * [`main`/src/container/tabs.rs#L307](https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs?rgh-link-date=2024-05-31T10%3A45%3A05Z#L307)
> 
> As a result, the mouse cursor remains the drag hand even when hovering the close button.

If you want to have a normal cursor when hovering the close-button, try this:

```diff
--- a/src/behavior.rs
+++ b/src/behavior.rs
@@ -105,7 +105,9 @@ pub trait Behavior<Pane> {
             + f32::from(state.closable) * (close_btn_left_padding + close_btn_size.x);
         let (_, tab_rect) = ui.allocate_space(vec2(button_width, ui.available_height()));
 
-        let tab_response = ui.interact(tab_rect, id, Sense::click_and_drag());
+        let tab_response = ui
+            .interact(tab_rect, id, Sense::click_and_drag())
+            .on_hover_cursor(egui::CursorIcon::Grab);
 
         // Show a gap when dragged
         if ui.is_rect_visible(tab_rect) && !state.is_being_dragged {
@@ -139,7 +141,8 @@ pub trait Behavior<Pane> {
 
                 // Allocate
                 let close_btn_id = ui.auto_id_with("tab_close_btn");
-                let close_btn_response = ui.interact(close_btn_rect, close_btn_id, Sense::click());
+                let close_btn_response =
+                    ui.interact(close_btn_rect, close_btn_id, Sense::click_and_drag());
 
                 let visuals = ui.style().interact(&close_btn_response);
 
diff --git a/src/container/tabs.rs b/src/container/tabs.rs
index 13455fc..e88e49c 100644
--- a/src/container/tabs.rs
+++ b/src/container/tabs.rs
@@ -303,7 +303,6 @@ impl Tabs {
 
                             let response =
                                 behavior.tab_ui(&mut tree.tiles, ui, id, child_id, tab_state);
-                            let response = response.on_hover_cursor(egui::CursorIcon::Grab);
                             if response.clicked() {
                                 behavior.on_edit(EditAction::TabSelected);
                                 next_active = Some(child_id);
```

This also forbids dragging the close-button, for better or for worse

File: src/tiles.rs [hunk 18]
```diff
@@ -74,6 +74,15 @@ impl<Pane> Tiles<Pane> {
         self.tiles.get(&tile_id)
     }
 
+    /// Get the pane instance for a given [`TileId`]
+    #[allow(dead_code)]
+    pub fn get_pane(&self, tile_id: &TileId) -> Option<&Pane> {
```
Comment (emilk) [inline src/tiles.rs:79]: I suggest adding `get_container` too for completeness

File: src/tiles.rs [hunk 19]
```diff
@@ -74,6 +74,15 @@ impl<Pane> Tiles<Pane> {
         self.tiles.get(&tile_id)
     }
 
+    /// Get the pane instance for a given [`TileId`]
+    #[allow(dead_code)]
```
Comment (emilk) [inline src/tiles.rs:78]: shouldn't be needed if this is part of the public API

```suggestion
```

File: src/behavior.rs [hunk 20]
```diff
@@ -22,17 +22,45 @@ pub enum EditAction {
...
+    /// Called when the close-button on a tab is pressed.
+    ///
+    /// Return `false` to abort the closing of a tab (e.g. after showing a message box).
+    fn on_tab_close(&mut self, _tiles: &mut Tiles<Pane>, _tile_id: TileId) -> bool {
+        true
+    }
+
+    /// The size of the close button in the tab
+    fn close_button_size(&self) -> f32 {
+        12.0
```
[the remark this answers is not in the document]
Comment (voidburn) [reply src/behavior.rs:61]: Yeah this would be better indeed. I'll get to it asap!

**Comment by voidburn** [comment]
> > P.S. In my personal implementation I commented out all `.on_hover_cursor(egui::CursorIcon::Grab)` except for the one that changes the cursor while dragging, as they were overriding all my attempts to change the cursor to a pointer when over the close button. I don't know how that would've impacted existing implementations so I left them in this PR.
> > Cursor changes in question:
> > 
> > * [`main`/src/container/tabs.rs#L279](https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs?rgh-link-date=2024-05-31T10%3A45%3A05Z#L279)
> > * [`main`/src/container/tabs.rs#L307](https://github.com/rerun-io/egui_tiles/blob/main/src/container/tabs.rs?rgh-link-date=2024-05-31T10%3A45%3A05Z#L307)
> > 
> > As a result, the mouse cursor remains the drag hand even when hovering the close button.
> 
> If you want to have a normal cursor when hovering the close-button, try this:
> 
> ```diff
> --- a/src/behavior.rs
> +++ b/src/behavior.rs
> @@ -105,7 +105,9 @@ pub trait Behavior<Pane> {
>              + f32::from(state.closable) * (close_btn_left_padding + close_btn_size.x);
>          let (_, tab_rect) = ui.allocate_space(vec2(button_width, ui.available_height()));
>  
> -        let tab_response = ui.interact(tab_rect, id, Sense::click_and_drag());
> +        let tab_response = ui
> +            .interact(tab_rect, id, Sense::click_and_drag())
> +            .on_hover_cursor(egui::CursorIcon::Grab);
>  
>          // Show a gap when dragged
>          if ui.is_rect_visible(tab_rect) && !state.is_being_dragged {
> @@ -139,7 +141,8 @@ pub trait Behavior<Pane> {
>  
>                  // Allocate
>                  let close_btn_id = ui.auto_id_with("tab_close_btn");
> -                let close_btn_response = ui.interact(close_btn_rect, close_btn_id, Sense::click());
> +                let close_btn_response =
> +                    ui.interact(close_btn_rect, close_btn_id, Sense::click_and_drag());
>  
>                  let visuals = ui.style().interact(&close_btn_response);
>  
> diff --git a/src/container/tabs.rs b/src/container/tabs.rs
> index 13455fc..e88e49c 100644
> --- a/src/container/tabs.rs
> +++ b/src/container/tabs.rs
> @@ -303,7 +303,6 @@ impl Tabs {
>  
>                              let response =
>                                  behavior.tab_ui(&mut tree.tiles, ui, id, child_id, tab_state);
> -                            let response = response.on_hover_cursor(egui::CursorIcon::Grab);
>                              if response.clicked() {
>                                  behavior.on_edit(EditAction::TabSelected);
>                                  next_active = Some(child_id);
> ```
> 
> This also forbids dragging the close-button, for better or for worse

I added a way to customize it, let me know if you want it to default to something different. I think everything looks good now!

File: src/behavior.rs [hunk 21]
```diff
@@ -113,7 +118,9 @@ pub trait Behavior<Pane> {
             + f32::from(state.closable) * (close_btn_left_padding + close_btn_size.x);
         let (_, tab_rect) = ui.allocate_space(vec2(button_width, ui.available_height()));
 
-        let tab_response = ui.interact(tab_rect, id, Sense::click_and_drag());
+        let tab_response = ui
+            .interact(tab_rect, id, Sense::click_and_drag())
+            .on_hover_cursor(egui::CursorIcon::Grab);
```
Comment (emilk) [inline src/behavior.rs:123]: If you add it here, remove it from the other place:

``` diff
--- a/src/container/tabs.rs
+++ b/src/container/tabs.rs
@@ -303,7 +303,6 @@ impl Tabs {
 
                             let response =
                                 behavior.tab_ui(&mut tree.tiles, ui, id, child_id, tab_state);
-                            let response = response.on_hover_cursor(egui::CursorIcon::Grab);
                             if response.clicked() {
                                 behavior.on_edit(EditAction::TabSelected);
                                 next_active = Some(child_id);
```

File: src/behavior.rs [hunk 22]
```diff
@@ -147,7 +154,9 @@ pub trait Behavior<Pane> {
 
                 // Allocate
                 let close_btn_id = ui.auto_id_with("tab_close_btn");
-                let close_btn_response = ui.interact(close_btn_rect, close_btn_id, Sense::click());
+                let close_btn_response = ui
+                    .interact(close_btn_rect, close_btn_id, Sense::click_and_drag())
+                    .on_hover_cursor(self.close_button_hover_cursor());
```
Comment (emilk) [inline src/behavior.rs:159]: note that due to button grabbing the `drag` sense, dragging the close button does nothing (which I think is a good thing), so the user changing the cursor for it to anything but the default would be quite confusing imho

**Comment by emilk** [comment]
> I think everything looks good now!

The CI disagrees

File: src/behavior.rs (hunk 22 above)
Comment (voidburn) [reply src/behavior.rs:159]: My reasoning about this was: 

1. On the web something clickable usually uses a hand pointing cursor.
2. On desktop applications something clickable does not usually affect the cursor (rather leverages other visual feedback mechanisms)

I though that if someone were to make an application with egui & egui_tiles and intends to only publish it as a web application they might welcome having the option to customize it.

What do you think?

Comment (emilk) [reply src/behavior.rs:159]: An egui user can already change the default cursor for _all_ buttons using https://docs.rs/egui/latest/egui/style/struct.Visuals.html#structfield.interact_cursor

Comment (voidburn) [reply src/behavior.rs]: I had completely forgotten about that! Thank you. Removing it.

**Comment by emilk** [comment]
Thanks!

**Comment by voidburn** [comment]
Thank you for Egui and all these goodies, happy to help!

**Pull request merged**